openraft/src/replication/mod.rs

```diff
@@ -568,26 +564,41 @@ impl<C: RaftTypeConfig, N: RaftNetworkFactory<C>, S: RaftStorage<C>> Replication
             };
 
             // TODO(xp): the span is not used. remove it from event.
-            self.process_raft_event(ev_and_span.0)?;
+            self.process_raft_event(ev_and_span.0)?
         }
 
         Ok(())
     }
 
+    /// return true if commit index update, in this way, MUST `send_append_entries` to sync new commit index
+    fn check_if_update_commit_index(&mut self, new_committed_index: Option<LogId<C::NodeId>>) -> bool {
+        assert!(new_committed_index >= self.committed);
+        if new_committed_index > self.committed {
+            self.committed = new_committed_index;
+            return true;
+        }
+        false
+    }
+
     #[tracing::instrument(level = "trace", skip(self), fields(event=%event.summary()))]
     pub fn process_raft_event(&mut self, event: RaftEvent<C>) -> Result<(), ReplicationError<C>> {
         tracing::debug!(event=%event.summary(), "process_raft_event");
 
         match event {
             RaftEvent::UpdateCommittedLogId { committed } => {
-                self.committed = committed;
+                if self.check_if_update_commit_index(committed) {
+                    self.need_to_replicate = true;
+                }
```

Comment:
Maybe it is more clear without `check_if_update_commit_index()`:
```suggestion
                self.need_to_replicate = committed > self.committted
                self.committed = committed;
```